src/bin/download_sysext.rs

```diff
@@ -113,20 +135,36 @@ impl<'a> Package<'a> {
         }
     }
 
-    fn verify_signature_on_disk(&mut self, from_path: &Path, pubkey_path: &str) -> Result<(), Box<dyn Error>> {
+    fn verify_signature_on_disk(&mut self, from_path: &Path, pubkey_path: &str) -> Result<Vec<u8>, Box<dyn Error>> {
         let upfile = File::open(from_path)?;
 
+        // create a BufReader to pass down to parsing functions.
+        let upfreader = &mut BufReader::new(upfile);
+
         // Read update payload from file, read delta update header from the payload.
-        let res_data = fs::read_to_string(from_path);
+        let header = delta_update::read_delta_update_header(upfreader)?;
 
-        let header = delta_update::read_delta_update_header(&upfile)?;
+        let mut delta_archive_manifest = delta_update::get_manifest_bytes(upfreader, &header)?;
 
         // Extract signature from header.
-        let sigbytes = delta_update::get_signatures_bytes(&upfile, &header)?;
+        let sigbytes = delta_update::get_signatures_bytes(upfreader, &header, &mut delta_archive_manifest)?;
 
-        // Parse signature data from the signature containing data, version, special fields.
-        let _sigdata = match delta_update::parse_signature_data(res_data.unwrap().as_bytes(), &sigbytes, pubkey_path) {
-            Some(data) => data,
+        // Extract header and data blobs together.
+        let header_data = delta_update::get_header_data(upfreader, &header, &delta_archive_manifest)?;
+
+        // Extract data blobs
+        let datablobs = delta_update::get_data_blobs(upfreader, &header, &delta_archive_manifest)?;
+
+        // Check for hash of data blobs with new_partition_info hash.
+        let pinfo_hash = delta_archive_manifest.new_partition_info.hash.clone().unwrap_or_default();
```

Comment:
The use of or_default means that it won't match the one we compute below.
Maybe annotate that this here means we enforce having a hash in `new_partition_info` and not having one will fail. Or do an explicit check with `if let` where we return an error in `else`.

Reply:
Done with error checking